pycommand.3
.\" Man page generated from reStructuredText.
.
.TH "PYCOMMAND" "3" "June 04, 2015" "0.3.0" "pycommand"
.SH NAME
pycommand \- pycommand Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
\fI\%Build Status\fP
.sp
\fBA clean and simplistic alternative for argparse, optparse and getopt\fP
.sp
Pycommand is essentially a fancy wrapper around getopt that consists of
one simple \fICommandBase\fP class that you can use to create executable
commands for your python programs with very simplistic and readable
code. It has support for subcommands and also nesting commands, so you
can create (multiple levels of) subcommands, with the ability to pass
the values of optional arguments of a command object to its subcommand
objects. Supported Python versions are 2.7 and 3.2 and later.
.INDENT 0.0
.IP \(bu 2
Github: \fI\%https://github.com/babab/pycommand\fP
.IP \(bu 2
Bitbucket: \fI\%https://bitbucket.org/babab/pycommand\fP
.IP \(bu 2
PyPI: \fI\%https://pypi.python.org/pypi/pycommand/\fP
.IP \(bu 2
Documentation: \fI\%http://pythonhosted.org/pycommand/\fP
.UNINDENT
.SH DOWNLOAD AND INSTALL
.sp
If you have pip installed, you can just:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
# pip install pycommand
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
To work with the current development version, do something like this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ git clone git://bitbucket.org/babab/pycommand.git
# cd pycommand
# pip install \-e .
.ft P
.fi
.UNINDENT
.UNINDENT
.SH QUICK SETUP FROM A TEMPLATE
.sp
To quicly start writing a command from a template (much like the
examples below), use pycommand\(aqs helper script by running:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ pycommand init
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This will ask you for an executable name, class name and template type
and it will save it to an executable file, ready to be used as a Python
shell script (for your Python package/module)).
.SH EXAMPLE #1 - A BASIC COMMAND
.sp
Here is a demonstration of the automated usage text generation and
parsing of optional arguments.
.sp
If we name the script for which you can see the code below
\fBbasic\-example\fP and execute it, the following will be the output for
running \fBbasic\-example \-h\fP or \fBbasic\-example \-\-help\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: basic\-example [options]

An example of a basic CLI program

Options:
\-h, \-\-help                        show this help information
\-f <filename>, \-\-file=<filename>  use specified file
\-\-version                         show version information
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
And here is the code:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
#!/usr/bin/env python

import pycommand
import sys


class BasicExampleCommand(pycommand.CommandBase):
    \(aq\(aq\(aqAn example of a basic CLI program

    This is an example that demonstrates the handling of all possible
    settings for the optional arguments, managed via the \(gaoptionList\(ga.
    This example does not handle positional arguments.

    \(aq\(aq\(aq
    usagestr = \(aqusage: basic\-example [options]\(aq
    description = __doc__.split(\(aq\en\(aq)[0]

    # optionList is a tuple of 2\-tuples, in format:
    # (long\-option, (short\-option, argument, help\-information))
    #
    # The order in which you define the options will be the order
    # in which they will appear in the usage message
    optionList = (
        (\(aqhelp\(aq, (\(aqh\(aq, False, \(aqshow this help information\(aq)),

        # To specify that an option requires an argument just add a
        # string that describes it
        (\(aqfile\(aq, (\(aqf\(aq, \(aq<filename>\(aq, \(aquse specified file\(aq)),

        # Use an empty string to ommit short option. Long option names
        # cannot be ommitted, since they are used as dictionary keys in
        # \(gaself.flags\(ga which holds the validated input
        (\(aqversion\(aq, (\(aq\(aq, False, \(aqshow version information\(aq)),
    )

    def run(self):
        \(aq\(aq\(aqThe \(garun\(ga method of the main command

        You need to define a method in your class that actually deals
        with any options that the user of your program has set. We call
        it \(garun\(ga here, but you can name it whatever you want.

        After the object has been created, there are 4 instance
        variables ready for you to use to write the flow of the program.
        In this example we only use the following three::

            error \-\- Thrown by GetoptError when parsing illegal
                     arguments

            flags \-\- Object/dict of parsed options and corresponding
                     arguments, if any.

            usage \-\- String with usage information. The string
                     is compiled using the values found for \(gausagestr\(ga,
                     \(gadescription\(ga, \(gaoptionList\(ga and \(gausageTextExtra\(ga.

        \(aq\(aq\(aq
        if self.flags.help:
            print(self.usage)
            return 0
        elif self.flags.version:
            print(\(aqPython version \(aq + sys.version.split()[0])
            return 0
        elif self.flags.file:
            print(\(aqfilename = \(aq + self.flags.file)
            return 0

        print(\(aqProgram completed. Try adding "\-\-help"\(aq)

if __name__ == \(aq__main__\(aq:
    # Shortcut for reading from sys.argv[1:] and sys.exit(status)
    pycommand.run_and_exit(BasicExampleCommand)

    # The shortcut is equivalent to the following:

    # cmd = BasicExampleCommand(sys.argv[1:])
    # if cmd.error:
    #     print(\(aqerror: {0}\(aq.format(cmd.error))
    #     sys.exit(1)
    # else:
    #     sys.exit(cmd.run())
.ft P
.fi
.UNINDENT
.UNINDENT
.SH EXAMPLE #2 - FULL EXAMPLE OF ONE MAIN COMMAND WITH TWO SUBCOMMANDS
.sp
Here is a full example demonstrating essentially the same program, but
with the \fB\-\-help\fP and \fB\-\-version\fP options replaced for subcommands:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
#!/usr/bin/env python

import pycommand
import sys


class VersionCommand(pycommand.CommandBase):
    usagestr = \(aqusage: full\-example version\(aq
    description = \(aqShow version information\(aq

    def run(self):
        print(\(aqPython version \(aq + sys.version.split()[0])
        print(\(aqFileflag = {0}\(aq.format(self.parentFlags[\(aqfile\(aq]))


class HelpCommand(pycommand.CommandBase):
    usagestr = \(aqusage: full\-example help [<command>]\(aq
    description = \(aqShow help information\(aq

    def run(self):
        if self.args and self.args[0] == \(aqversion\(aq:
            print(VersionCommand([]).usage)
        print(cmd.usage)


class FullExampleCommand(pycommand.CommandBase):
    \(aq\(aq\(aqAn full example of a pycommand CLI program

    This is an example that demonstrates the mapping of subcommands
    and registrering the \-\-file flag from the main command to its
    subcommand. It only explains new concepts that are not handled in
    \(ga\(gabasic\-example\(ga\(ga, so be sure to see that first.

    \(aq\(aq\(aq
    usagestr = \(aqusage: full\-example [\-f <filename>] <command> [<args>]\(aq
    description = (
        \(aqCommands:\en\(aq
        \(aq   help         show this help information\en\(aq
        \(aq   version      show full version information\(aq
    )

    # Mapping of subcommands
    commands = {\(aqhelp\(aq: HelpCommand,
                \(aqversion\(aq: VersionCommand}

    optionList = ((\(aqfile\(aq, (\(aqf\(aq, \(aq<filename>\(aq, \(aquse specified file\(aq)), )

    # Optional extra usage information
    usageTextExtra = (
        "See \(aqfull\-example help <command>\(aq for more information on a "
        "specific command."
    )

    def run(self):
        \(aq\(aq\(aqThe \(garun\(ga method of the main command

        After the object has been created, there are 4 instance
        variables ready for you to use to write the flow of the program.
        In this example we use them all::

            error \-\- Thrown by GetoptError when parsing illegal
                     arguments

            flags \-\- OrderedDict of parsed options and corresponding
                     arguments, if any.

            usage \-\- String with usage information. The string
                     is compiled using the values found for \(gausagestr\(ga,
                     \(gadescription\(ga, \(gaoptionList\(ga and \(gausageTextExtra\(ga.

            parentFlags \-\- Dict of registered \(gaflags\(ga of another
                           \(gaCommandBase\(ga object.

        \(aq\(aq\(aq
        try:
            cmd = super(FullExampleCommand, self).run()
        except pycommand.CommandExit as e:
            return e.err

        # Register a flag of a parent command
        # :Parameters:
        #     \- \(gaoptionName\(ga: String. Name of option
        #     \- \(gavalue\(ga: Mixed. Value of parsed flag\(ga
        cmd.registerParentFlag(\(aqfile\(aq, self.flags.file)

        if cmd.error:
            print(\(aqfull\-example {cmd}: {error}\(aq
                  .format(cmd=self.args[0], error=cmd.error))
            return 1
        else:
            return cmd.run()


if __name__ == \(aq__main__\(aq:
    # Shortcut for reading from sys.argv[1:] and sys.exit(status)
    pycommand.run_and_exit(FullExampleCommand)
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
And here are some output examples:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ ./full\-example
usage: full\-example [\-f <filename>] <command> [<args>]

Commands:
   help         show this help information
   version      show full version information

Options:
\-f <filename>, \-\-file=<filename>  use specified file

See \(aqfull\-example help <command>\(aq for more information on a specific command.

$ ./full\-example help version
usage: full\-example version

Show version information

$ ./full\-example \-f
error: option \-f requires argument

$ ./full\-example \-f somefilename version
Python version 3.3.2
Fileflag = somefilename

$ ./full\-example version
Python version 3.3.2
Fileflag = None

$ ./full\-example help doesnotexist
error: command doesnotexist does not exist
.ft P
.fi
.UNINDENT
.UNINDENT
.SH MODULE DOCUMENTATION
.sp
A clean and simplistic alternative for argparse, optparse and getopt
.INDENT 0.0
.TP
.B class pycommand.CommandBase(argv=[\(aq\-b\(aq, \(aqman\(aq, \(aq\-d\(aq, \(aq_build/doctrees\(aq, \(aq.\(aq, \(aq_build/man\(aq])
Base class for (sub)commands
.INDENT 7.0
.TP
.B args = None
List of parsed postional arguments
.UNINDENT
.INDENT 7.0
.TP
.B commands = {}
Dictionary of commands and the callables they invoke.
.UNINDENT
.INDENT 7.0
.TP
.B description = \(aq\(aq
String. Small description of subcommand
.UNINDENT
.INDENT 7.0
.TP
.B error = None
Thrown by GetoptError when parsing illegal arguments.
.UNINDENT
.INDENT 7.0
.TP
.B flags = None
Dict of parsed options and corresponding arguments, if any.
.UNINDENT
.INDENT 7.0
.TP
.B optionList = {}
Dictionary of options (as a tuple of 2\-tuples).
This will be transformed to an OrderedDict when initializing the object.
.sp
Example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
optionList = (
    (\(aqhelp\(aq, (\(aqh\(aq, False, \(aqshow this help information\(aq)),
    (\(aqdry\-run\(aq, (\(aqn\(aq, False,
                 \(aqonly print output without actually running\(aq)),

    # To specify that an option requires an argument
    # just add a string that describes it
    (\(aqfile\(aq, (\(aqf\(aq, \(aq<filename>\(aq, \(aquse specified file\(aq)),

    # Use an empty string to ommit short option
    (\(aqdebug\(aq, (\(aq\(aq, False, \(aqshow debug information\(aq)),
)
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B parentFlags = None
Dict of registered \fIflags\fP of parent Command object.
.UNINDENT
.INDENT 7.0
.TP
.B registerParentFlag(optionName, value)
Register a flag of a parent command
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fIoptionName\fP: String. Name of option
.IP \(bu 2
\fIvalue\fP: Mixed. Value of parsed flag\(ga
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B usage = None
String with usage information
.sp
The string is compiled using the values found for
\fIusagestr\fP, \fIdescription\fP, \fIoptionList\fP and \fIusageTextExtra\fP\&.
.UNINDENT
.INDENT 7.0
.TP
.B usageTextExtra = \(aq\(aq
String. Optional extra usage information
.UNINDENT
.INDENT 7.0
.TP
.B usagestr = \(aqusage: command [options]\(aq
String. Usage synopsis
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B exception pycommand.OptionError
Options/Flags AttributeError exception
.UNINDENT
.INDENT 0.0
.TP
.B class pycommand.dictobject
A dictionary with getters by attribute, used for flags
.UNINDENT
.INDENT 0.0
.TP
.B pycommand.run_and_exit(command_class)
A shortcut for reading from sys.argv and exiting the interpreter
.UNINDENT
.SH WHY WAS IT CREATED?
.sp
When parsing command line program arguments, I sometimes work with
\fIargparse\fP (a replacement for \fIoptparse\fP). I don\(aqt really like the API
and the output it gives, which is the main reason I\(aqve always used
\fIgetopt\fP for parsing arguments whenever possible.
.sp
The \fICommandBase\fP class was originally written for \fIDisPass\fP,
which is a password manager/generator, as a means to easily define new
subcommands and have auto\-generated usage messages. Because I want to
have this in other projects I\(aqve decided to put it in the cheeseshop.
.SH CHANGE LOG
.sp
pycommand adheres to \fI\%Semantic Versioning\fP\&.
.SS 0.3.0 \- 2015\-06\-04
.SS Added
.INDENT 0.0
.IP \(bu 2
Shortcut \fBrun_and_exit()\fP for reading from \fBsys.argv[1:]\fP and exiting
the interpreter via \fBsys.exit(status)\fP
.IP \(bu 2
Package as wheel distribution to speed up installations
.IP \(bu 2
Add \fBman pycommand\fP ability, i.e. install mandoc in \fB/usr/share/man3/\fP
.UNINDENT
.SS Changed
.INDENT 0.0
.IP \(bu 2
Add support for getting flags by attribute like \fBself.flags.help\fP\&.
The default approach for normal dicts like \fBself.flags[\(aqhelp\(aq]\fP
remains valid.
.UNINDENT
.SS 0.2.0 \- 2015\-05\-21
.SS Added
.INDENT 0.0
.IP \(bu 2
Full example of a command with subcommands
.IP \(bu 2
Create quick templates via pycommand script (\fBpycommand init\fP)
.IP \(bu 2
Unit tests and automatic testing via Travis\-CI
.IP \(bu 2
Documentation \fBman\fP (.3) and \fBinfo\fP (.info) pages
.UNINDENT
.SS Changed
.INDENT 0.0
.IP \(bu 2
Specification of subcommands can be \fI\%defined in CommandBase.command\fP
as a shortcut.
.UNINDENT
.SS 0.1.0 \- 2013\-08\-08
.SS Added
.INDENT 0.0
.IP \(bu 2
Initial release
.UNINDENT
.SH SOFTWARE LICENSE
.sp
Copyright (c) 2013\-2015  Benjamin Althues <\fI\%benjamin@babab.nl\fP>
.sp
Permission to use, copy, modify, and distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.
.sp
THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.SH AUTHOR
Benjamin Althues
.SH COPYRIGHT
2013-2015, Benjamin Althues
.\" Generated by docutils manpage writer.
.


CHANGELOG.md
# Change Log

All notable changes to pycommand will be documented here. The project
adheres to [Semantic Versioning](http://semver.org/).


## 0.0.2 - TO BE RELEASED
### Added
- Full example of a command with subcommands
- Create quick templates via pycommand script (pycommand init)
- Unit tests and automatic testing via Travis-CI
- Documentation `man` (.3) and `info` (.info) pages

### Changed
- Specification of subcommands can be
  [defined in CommandBase.command](https://github.com/babab/pycommand/commit/a978a05ef92f70f0ce6b06d96a38c2caa8297ecc)
  as a shortcut.


## 0.0.1 - 2003-08-08
### Added
- Initial release


pycommand.py
'''A clean and simplistic alternative for argparse, optparse and getopt'''

# Copyright (c) 2013-2015  Benjamin Althues <benjamin@babab.nl>
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

from collections import OrderedDict
import getopt
import sys

__docformat__ = 'restructuredtext'
__author__ = "Benjamin Althues"
__copyright__ = "Copyright (C) 2013-2015  Benjamin Althues"
__version_info__ = (0, 3, 0, 'beta', 0)
__version__ = '0.3.0'


class CommandExit(Exception):
    def __init__(self, val):
        self.err = val

    def __str__(self):
        return repr(self.err)


class OptionError(AttributeError):
    '''Options/Flags AttributeError exception'''


class dictobject(dict):
    '''A dictionary with getters by attribute, used for flags '''
    def __getattr__(self, name):
        if name in self:
            return self[name]
        else:
            raise OptionError("Option '{}' is not defined".format(name))


class CommandBase(object):
    '''Base class for (sub)commands'''

    usagestr = 'usage: command [options]'
    '''String. Usage synopsis'''

    description = ''
    '''String. Small description of subcommand'''

    optionList = {}
    '''Dictionary of options (as a tuple of 2-tuples).
    This will be transformed to an OrderedDict when initializing the object.

    Example::

        optionList = (
            ('help', ('h', False, 'show this help information')),
            ('dry-run', ('n', False,
                         'only print output without actually running')),

            # To specify that an option requires an argument
            # just add a string that describes it
            ('file', ('f', '<filename>', 'use specified file')),

            # Use an empty string to ommit short option
            ('debug', ('', False, 'show debug information')),
        )

    '''

    usageTextExtra = ''
    '''String. Optional extra usage information'''

    commands = {}
    '''Dictionary of commands and the callables they invoke.'''

    def __init__(self, argv=sys.argv[1:]):
        '''Initialize (sub)command object

        :Parameters:
            - `argv`: List of arguments. E.g. `sys.argv[1:]`
        '''

        # Instance vars
        self.error = None
        '''Thrown by GetoptError when parsing illegal arguments.'''

        self.flags = {}
        '''Dict of parsed options and corresponding arguments, if any.'''

        self.args = []
        '''List of parsed postional arguments'''

        self.parentFlags = {}
        '''Dict of registered `flags` of parent Command object.'''

        self.usage = ''
        '''String with usage information

        The string is compiled using the values found for
        `usagestr`, `description`, `optionList` and `usageTextExtra`.
        '''

        self.optionList = OrderedDict(self.optionList)

        # Local vars
        longopts = []
        padding = 0
        shortopts = ''

        # Calculate padding needed for option arguments in usage info
        for flag, val in self.optionList.items():
            optlen = len(flag) + 2
            optlen += 4 if val[0] else 0
            optlen += len(val[1]) + 1 if val[0] and val[1] else 0
            optlen += len(val[1]) + 1 if val[1] else 0
            padding = optlen if optlen > padding else padding

        # Create usage information and build dict of possible flags
        opthelp = ''
        for flag, val in self.optionList.items():
            spec = flag + '=' if val[1] else flag
            longopts.append(spec)
            self.flags.update({flag: None})

            if val[1]:
                flagstring_long = ('{flag}={argument}'
                                   .format(flag=flag, argument=val[1]))
                if val[0]:
                    flagstring_short = ('{flag} {argument}'
                                        .format(flag=val[0], argument=val[1]))
            else:
                flagstring_long = flag
                flagstring_short = val[0]

            if val[0]:
                shortopts += val[0] + ':' if val[1] else val[0]
                optline = ('-{short}, --{flag}'
                           .format(short=flagstring_short,
                                   flag=flagstring_long))
            else:
                optline = '--{flag}'.format(flag=flagstring_long)

            opthelp += ('{options:{padding}}  {desc}\n'
                        .format(options=optline, padding=padding, desc=val[2]))

        self.usage = self.usagestr
        if self.description:
            self.usage += '\n\n{desc}'.format(desc=self.description)
        if self.optionList:
            self.usage += '\n\nOptions:\n{opts}'.format(opts=opthelp)
        if self.usageTextExtra:
            self.usage += '\n{help}'.format(help=self.usageTextExtra)

        # Parse arguments and options
        try:
            opts, self.args = getopt.getopt(argv, shortopts, longopts)
        except getopt.GetoptError as err:
            self.error = err
            return  # Stop when an invalid option is parsed

        for opt in opts:
            # Compare each option with optionList and set values for flags
            for flag, val in self.optionList.items():
                if opt[0][1] != '-':
                    # Short tags
                    if opt[0][1] == val[0]:
                        if val[1]:
                            self.flags[flag] = opt[1]
                        else:
                            self.flags[flag] = True
                else:
                    # Long tags
                    if opt[0][2:] == flag:
                        if val[1]:
                            self.flags[flag] = opt[1]
                        else:
                            self.flags[flag] = True

        # Convert to dictobject to allow getting flags by attribute name
        self.flags = dictobject(self.flags)

    def run(self):
        if not self.args:
            print(self.usage)
            raise CommandExit(2)
        elif self.args[0] in self.commands:
            return self.commands[self.args[0]](argv=self.args[1:])
        else:
            print('error: command {cmd} does not exist'
                  .format(cmd=self.args[0]))
            raise CommandExit(1)

    def registerParentFlag(self, optionName, value):
        '''Register a flag of a parent command

        :Parameters:
            - `optionName`: String. Name of option
            - `value`: Mixed. Value of parsed flag`
        '''
        self.parentFlags.update({optionName: value})
        return self


def run_and_exit(command_class):
    '''A shortcut for reading from sys.argv and exiting the interpreter'''
    cmd = command_class(sys.argv[1:])
    if cmd.error:
        print('error: {0}'.format(cmd.error))
        sys.exit(1)
    else:
        sys.exit(cmd.run())
